Fix version number and table creation if not exists

nilskre opened the pull request:
This PR contains two fixes:
- In case the defined table does not exist it should be created. Currently, the executed query to check if the database exists already fails, because the not existing database name is part of the query. This PR fixes this behavior to create the table if it does not exist. 
- Fix to get the version number if `neo4j-admin` is not present (e.g. in a CICD setup with Neo4j in a Docker container).

@deeenes Can you have a look at it?
Thanks in advance!

nilskre committed:
fix type error in query

nilskre committed:
make it compatible with python versions <3.10

nilskre committed:
pass on db name

nilskre committed:
_driver has method to get the neo4j version

nilskre committed:
_driver has method to get the neo4j version

nilskre committed:
correctly handle the case when the db table does not exist

nilskre committed:
fix check if db exists

nilskre committed:
changes with fallback_db broke the function to create the db if not exists -> bring back this functionality

nilskre committed:
dont rely on admin-import being present (fails if neo4j is running e.g. in a docker container and not on the host); get it directly from the db

nilskre committed:
return major version directly

deeenes commented on neo4j_utils/_query.py:
```diff
@@ -29,4 +31,4 @@ class Query(
 
 Query.__new__.__defaults__ = (None,)
-Query.__annotations__ = {'query': str, 'args': dict | None}
+Query.__annotations__ = {'query': str, 'args': Optional[dict]}
```
hm, unfortunately we have to support Py 3.8 for half more year, but this could be nicer with `typing.NamedTuple`, instead of `Optional` ([see example here](https://github.com/saezlab/pypath/blob/master/pypath/inputs/disgenet/_records.py))

nilskre replied:
As far as I know, the `X|Y` syntax was introduced in Python 3.10 (see [PEP 604](https://peps.python.org/pep-0604/)). The fix from my side was to make it downward compatible with older Python versions. `Optional` was introduced in Python 3.5 (see [PEP 484](https://peps.python.org/pep-0484/))

deeenes commented on neo4j_utils/_n4jversion.py:
```diff
@@ -11,7 +12,7 @@ class Neo4jVersion:
     """
 
-    major: int | None
+    major: Union[int, None]
```
bc this is a class attribute, I would expect `from __future__ import annotations` to work

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -1603,3 +1603,17 @@ def last_fail(self):
 
         return self._queries.get('last_failed')
+
+    def get_neo4j_version(self):
```
maybe remove the `get_` prefix and make it a `@property`

nilskre replied:
What would be the advantage of having a property instead?

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -566,12 +566,12 @@ def query(
-
-        db = db or self._db_config['db'] or neo4j.DEFAULT_DATABASE
+        if 'CREATE' not in query and 'SHOW DATABASES' not in query:
+            db = db or self._db_config['db'] or neo4j.DEFAULT_DATABASE
```
and what happens in case of `CREATE` or `SHOW` queries?

nilskre replied:
In the case of `CREATE` a new database is created. So, the query isn't run on a specified database. 
Same applies to `SHOW DATABASES`.
Moreover, if in these two cases the db is set (like in the version before), the queries fail, because it tries to execute the queries on a database that does not exist.

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -566,12 +566,12 @@ def query(
         raise_errors = (
             self._db_config['raise_errors']
-                if raise_errors is None else
+            if raise_errors is None else
```
This line was indented on purpose :)

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -834,7 +834,10 @@ def db_status(
         with self.fallback():
 
-            resp, summary = self.query(query=query, db=name)
+            try:
```
I think this error handling shouldn't happen here, the `query` method already takes care of it, doesn't it?

nilskre replied:
In the latest commit of this PR, there is no try except here any more [see](https://github.com/saezlab/neo4j-utils/blob/f08467d6d190d306a1bbbbbc823c5c3ffd946d5a/neo4j_utils/_driver.py#L835)

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -830,18 +830,17 @@ def db_status(
         name = name or self.current_db
 
-        query = f'SHOW DATABASES WHERE name = "{name}";'
+        query = 'SHOW DATABASES'
```
we want to check the status of one specific database, why not to use `WHERE`?

nilskre replied:
In the case that the database we are looking does not exist, this query fails.

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -830,18 +830,17 @@ def db_status(
-
-        if resp:
+            resp, summary = self.query(query=query)
+            databases = [record['name'] for record in resp]
```
why list the names of all the databases, when we're wondering about only one?

nilskre replied:
Related to the comment above ("In the case that the database we are looking does not exist, this query fails.")

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -830,18 +830,17 @@ def db_status(
+            databases = [record['name'] for record in resp]
 
+        if name in databases:
             return resp[0].get(field, resp[0])
```
I'm not getting how this makes sense: nothing guarantees that the element at index `0` will be the database with `name`

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -830,18 +830,17 @@ def db_status(
+            databases = [record['name'] for record in resp]
 
+        if name in databases:
             return resp[0].get(field, resp[0])
```
I'm not getting how this makes sense: nothing guarantees that the element at index `0` will be the database with `name`

nilskre replied:
True. I have adapted it accordingly

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -830,18 +830,17 @@ def db_status(
 
+        if name in databases:
             return resp[0].get(field, resp[0])
+        else:
```
but these 2 last lines are redundant in any case

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -1566,9 +1565,9 @@ def str_to_exc(e):
             return (
                 e.__class__
-                    if isinstance(e, Exception) else
+                if isinstance(e, Exception) else
```
in multi-line `if ... else` statements I indent the `if ... else` lines for better readability

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -1577,15 +1576,15 @@ def str_to_exc(e):
-                    issubclass(error, e)
-                    for e in errors
-                    if isinstance(e, type)
+                error in errors or
```
these 7 lines got an extra indentation

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -833,8 +833,7 @@ def db_status(
         with self.fallback():
-
-            resp, summary = self.query(query=query)
+            resp, summary = self.query(query=query, db=neo4j.DEFAULT_DATABASE)
```
the query is executed against the `DEFAULT_DATABASE` by `query` if no other option left, isn't it? why do we have to hardcode it here? what if `DEFAULT_DATABASE` is not available for any reason?

nilskre replied:
removed the assignment to the default db

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -1010,7 +1009,7 @@ def drop_indices_constraints(self):
         """
 
-        neo4j_version = _n4jversion.Neo4jVersion()
+        neo4j_version = self.get_neo4j_version()
```
this is indeed much better -- so the version is not necessary before connection?

deeenes commented on neo4j_utils/_driver.py:
```diff
@@ -1604,9 +1604,9 @@ def last_fail(self):
         return self._queries.get('last_failed')
 
-    def get_neo4j_version(self):
+    def get_neo4j_major_version(self) -> int:
```
This could be called each time the instance connects to a new server, and accessed as a property, which could be a tuple of integers (who knows if in the future the minor version number will matter too)

nilskre replied:
Now adapted to have a version property

nilskre committed:
act upon feedback

nilskre committed:
act upon feedback 2

nilskre committed:
act upon feedback 3

nilskre committed:
neo4j_version as property

nilskre committed:
neo4j_version as property

nilskre committed:
act upon feedback

nilskre committed:
python <3.10 compatibility

nilskre commented:
Thanks for your feedback @deeenes ! I have now acted upon it (or added some questions to the comments).

nilskre committed:
init property

nilskre committed:
init property

nilskre commented:
Thanks for the feedback @deeenes . I have now adapted it accordingly.

nilskre commented:
Hi @deeenes ,
here a little reminder regarding this PR. As mentioned above I have addressed your feedback (or added some questions). Can you have a look at it? Thanks in advance!

deeenes commented:
Hi Nils, I'm not ready to merge this PR because I'm not convinced it implements the desired behaviour. Especially creating databases implicitly, without a command from the user, I'm not sure is a good idea. A test suite would be very helpful in order to evaluate the behaviour under a variety of conditions. Let me know if you would like to create some tests, or if I should do it (this latter takes longer I guess).

nilskre commented:
Totally agree, that tests are crucial for ensuring the right behavior.
Unfortunately, immediately I don't have the time to write a fully featured test suite for neo4j-utils (at the moment there are no tests at all).
As a solution for now I have rolled back the neo4j-utils version in BioCypher to version `0.0.7` (with this version everything works without any problem, starting from version `0.0.8` I have experienced the described issues).
Consequently, we can close this PR.

deeenes commented:
Alright, at some point me or someone else can write those tests. No need for complete test coverage, testing for the here relevant behaviours would be already beneficial. Comparing `0.0.7` and `0.0.8`, the main difference is the use of more than one fallback databases. And I think what might cause a different behaviour is that only `TransientError` triggers the fallback, so it in fact never happens, compromising the whole idea. This can be fixed easily.

The other point, the missing `neo4j-admin` command: is the `neo4j` command available? Because that is able to tell the version too.

I will later go through again this PR and cherry-pick anything that makes sense to include.
